x-pack/filebeat/input/awscloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,125 @@
+	logStreams           []string
+	logStreamPrefix      string
+	startTime            int64
+	endTime              int64
+	prevEndTime          int64
+	workerSem            *awscommon.Sem
+	log                  *logp.Logger
+	metrics              *inputMetrics
+	store                *statestore.Store
+	workersListingMap    *sync.Map
+	workersProcessingMap *sync.Map
+}
+
+func newCloudwatchPoller(log *logp.Logger, metrics *inputMetrics,
+	store *statestore.Store,
+	awsRegion string, apiSleep time.Duration,
+	numberOfWorkers int, logStreams []string, logStreamPrefix string) *cloudwatchPoller {
+	if metrics == nil {
+		metrics = newInputMetrics(monitoring.NewRegistry(), "")
+	}
+
+	return &cloudwatchPoller{
+		numberOfWorkers:      numberOfWorkers,
+		apiSleep:             apiSleep,
+		region:               awsRegion,
+		logStreams:           logStreams,
+		logStreamPrefix:      logStreamPrefix,
+		startTime:            int64(0),
+		endTime:              int64(0),
+		workerSem:            awscommon.NewSem(numberOfWorkers),
+		log:                  log,
+		metrics:              metrics,
+		store:                store,
+		workersListingMap:    new(sync.Map),
+		workersProcessingMap: new(sync.Map),
+	}
+}
+
+func (p *cloudwatchPoller) run(svc cloudwatchlogsiface.ClientAPI, logGroup string, startTime int64, endTime int64, logProcessor *logProcessor) {
+	err := p.getLogEventsFromCloudWatch(svc, logGroup, startTime, endTime, logProcessor)
+	if err != nil {
+		var err *awssdk.RequestCanceledError
+		if errors.As(err, &err) {
+			p.log.Error("getLogEventsFromCloudWatch failed with RequestCanceledError: ", err)
+		}
+		p.log.Error("getLogEventsFromCloudWatch failed: ", err)
+	}
+}
+
+// getLogEventsFromCloudWatch uses FilterLogEvents API to collect logs from CloudWatch
+func (p *cloudwatchPoller) getLogEventsFromCloudWatch(svc cloudwatchlogsiface.ClientAPI, logGroup string, startTime int64, endTime int64, logProcessor *logProcessor) error {
+	// construct FilterLogEventsInput
+	filterLogEventsInput := p.constructFilterLogEventsInput(startTime, endTime, logGroup)
+
+	// make API request
+	req := svc.FilterLogEventsRequest(filterLogEventsInput)
+	paginator := cloudwatchlogs.NewFilterLogEventsPaginator(req)
+	for paginator.Next(context.TODO()) {
+		page := paginator.CurrentPage()
+		p.metrics.apiCallsTotal.Inc()
+
+		logEvents := page.Events
+		p.metrics.logEventsReceivedTotal.Add(uint64(len(logEvents)))
+		p.log.Debugf("Processing #%v events", len(logEvents))
+		err := logProcessor.processLogEvents(logEvents, logGroup, p.region)
+		if err != nil {
+			err = errors.Wrap(err, "processLogEvents failed")
+			p.log.Error(err)
+		}
+	}
+
+	if err := paginator.Err(); err != nil {
+		return errors.Wrap(err, "error FilterLogEvents with Paginator")
+	}
+
+	// This sleep is to avoid hitting the FilterLogEvents API limit(5 transactions per second (TPS)/account/Region).
+	p.log.Debugf("sleeping for %v before making FilterLogEvents API call again", p.apiSleep)
```

Comment:
If the paginator is affected by the throttling, this should go inside the `paginator.Next()` for block, shouldn't it?

Otherwise I'm not sure how effective is the throttling here: this is executed after the FilterLogEvents API call, and the method itself it's called multiple times in a goroutine each.
In the end it might be that we are doing more than 5 TPS/account/region, we will be throttled for a few, and then sleep for 5x `p.apiSleep`

I would move to throttling to https://github.com/elastic/beats/pull/29695/files#diff-caf6618b123c7fe852f46f717af459389cfc03b351ca89d8263e6500574993e3R188

where we can have a context with timeout of 5sec and and a counter. when we hit 5 we let the goroutine exhaust and sleep

Reply:
if we move there I wonder if we need to expose the `api_sleep` at all: we could backpressure as much as we want correctly without asking the user to find the right value

Reply:
Good catch, the sleep definitely should be moved inside the for loop!

> we could backpressure as much as we want correctly without asking the user to find the right value

Let's assume the worst scenario, which is we are making this API call to the same account same region. Then only 5 calls can be made per second. The safe value for `api_sleep` in this case would be 200ms. 
But what if the user has more than one Filebeat running calling this API to the same account same region but different log stream names? In this case, we would still need the user to be able to adjust the `api_sleep` value manually.